R/data_modify.R

```diff
@@ -0,0 +1,245 @@
+  column_names <- colnames(data)
+
+  # we check for character vector of expressions, in which case
+  # "dots" should be unnamed
+  if (is.null(names(dots))) {
+    # if we have multiple strings, concatenate them to a character vector
+    # and put it into a list...
+    if (length(dots) > 1) {
+      if (all(vapply(dots, is.character, logical(1)))) {
+        dots <- list(unlist(dots))
+      } else {
+        insight::format_error("You cannot mix string and literal representation of expressions.")
+      }
+    }
+    # expression is given as character string, e.g.
+    # a <- "double_SepWidth = 2 * Sepal.Width"
+    # data_modify(iris, a)
+    # or as character vector, e.g.
+    # data_modify(iris, c("var_a = Sepal.Width / 10", "var_b = Sepal.Width * 10"))
+    character_symbol <- tryCatch(.dynEval(dots[[1]]), error = function(e) NULL)
+    # do we have a character vector? Then we can proceed
+    if (is.character(character_symbol)) {
+      dots <- lapply(character_symbol, function(s) {
+        # turn value from character vector into expression
+        str2lang(.dynEval(s))
+      })
+      names(dots) <- vapply(dots, function(n) insight::safe_deparse(n[[2]]), character(1))
+    }
+  }
+
+  for (i in seq_along(dots)) {
+    # iterate expressions for new variables
+    symbol <- dots[[i]]
+
+    # expression is given as character string in a variable, but named, e.g.
+    # a <- "2 * Sepal.Width"
+    # data_modify(iris, double_SepWidth = a)
+    # we reconstruct the symbol as if it were provided as literal expression.
+    # However, we need to check that we don't have a character vector,
+    # like: data_modify(iris, new_var = "a")
+    # this one should be recycled instead.
+    if (!is.character(symbol)) {
+      eval_symbol <- .dynEval(symbol, ifnotfound = NULL)
+      if (is.character(eval_symbol)) {
+        symbol <- str2lang(paste0(names(dots)[i], " = ", eval_symbol))
+      }
+    }
+
+    # finally, we can evaluate expression and get values for new variables
+    new_variable <- try(with(data, eval(symbol)), silent = TRUE)
+
+    # successful, or any errors, like misspelled variable name?
+    if (inherits(new_variable, "try-error")) {
+      # in which step did error happen?
+      step_number <- switch(as.character(i),
+        "1" = "the first expression",
+        "2" = "the second expression",
+        "3" = "the third expression",
+        paste("expression", i)
+      )
+      step_msg <- paste0("There was an error in ", step_number, ".")
+      # try to find out which variable was the cause for the error
+      error_msg <- attributes(new_variable)$condition$message
+      if (grepl("object '(.*)' not found", error_msg)) {
+        error_var <- gsub("object '(.*)' not found", "\\1", error_msg)
+        insight::format_error(
+          paste0(step_msg, " Variable \"", error_var, "\" was not found in the dataset."),
+          .misspelled_string(colnames(data), error_var, "Possibly misspelled?")
```

Comment:
The "object not found" message is not necessarily due to a missing column. For example maybe I want to do something like this:

``` r
library(datawizard)

x <- 10
head(mtcars) |> 
  data_modify(y = x + 1)
#>                    mpg cyl disp  hp drat    wt  qsec vs am gear carb  y
#> Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4 11
#> Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4 11
#> Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1 11
#> Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1 11
#> Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2 11
#> Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0    3    1 11
```

but `x` isn't defined so I get the error:

``` r
library(datawizard)

head(mtcars) |> 
  data_modify(y = x + 1)
#> Error: There was an error in the first expression. Variable "x" was not found
#>   in the dataset.
#>   Possibly misspelled?
```

and the error message is misleading. I think it can be removed (and it would simplify the code)

Reply:
We could add "was not found in the dataset or otherwise defined". I would prefer to keep this for the more common situation where you misspell variable names.

Reply:
Ok, but then I think `Variable "x" was not found in the dataset and in the environment.` is clearer